### ISSUE TITLE
Nemo file manager extension

### ISSUE THREAD

**Issue by snr-lab** [rabbitvcs/rabbitvcs#251]
Hi,

First of all, thank you for this nice project.
I was using nemo-rabbitvcs with Linux Mint 18.2. Recently I have upgraded to 19.1 and come to know that this extension is removed from cinnamon. Is there any plan making it available again?

**Comment by leoaltmann** [comment]
Hello, I'm having issues with this as well.  Running latest Linux Mint Cinnamon 19.1.  I tried installing nemo-rabbitvcs_3.6.0+sylvia manually but couldn't get anything new to show up in Nemo.  

I tried editing /usr/share/nemo-python/extensions/RabbitVCS.py to comment out the dependence on Nemo 3.0 (commented out line 70) but it then complained about syntax errors in an unrelated area (line 48).  

Thanks!

**Comment by moisesbites** [comment]
Hi,

Thank you for RabbitVCS. It's a great software.

I had the same issues in LM 19.1 with nemo. I tried using Nautilus plugin, but there is a error too. 
:-|

**Comment by jfabini** [comment]
Hi,
same issue here. Managed to get RabbitVCS with Nemo running on Linux Mint 19.0 using manual install but with LM 19.1 it does no longer work. 
Thanks

**Comment by jaheme127** [comment]
I have the same issue, here. Seems to be that Nemo moved to Python3 in the last version. RabbitVCS itself is based on Python2 - It seems to be a lot of work to make it runnable on Python 3 I guess...

Any comments from the devs? 

PS: The 3.0 in line 70 is still correct, since version 4.0 still uses the 3.0 namespace.

**Comment by adamplumb** [comment]
You might want to try the gtk3 branch which works with python3 and gtk3.  It should be in a pretty usable state.

**Comment by jaheme127** [comment]
Thank you. it works with the gtk3 branch on Linux Mint 19.1 and Nemo 4.0, by simply following the instructions (Use setup.py and copying the RabbitVCS.py)

**Comment by moisesbites** [comment]
I teste, and with root is working with gtk3 branch. Bur with my local user, it's not working.

https://github.com/rabbitvcs/rabbitvcs/issues/163

**Comment by moisesbites** [comment]
In the ~/.config/rabbitvcs/RabbitVCS.log has the log:
`
2019-04-03 12:43:40,279 ERROR	rabbitvcs.services.checkerservice	org.freedesktop.DBus.Error.ServiceUnknown: The name org.google.code.rabbitvcs.RabbitVCS.Checker was not provided by any .service files
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/rabbitvcs/services/checkerservice.py", line 262, in _connect_to_checker
    OBJECT_PATH)
  File "/usr/lib/python2.7/dist-packages/dbus/bus.py", line 241, in get_object
    follow_name_owner_changes=follow_name_owner_changes)
  File "/usr/lib/python2.7/dist-packages/dbus/proxies.py", line 248, in __init__
    self._named_service = conn.activate_name_owner(bus_name)
  File "/usr/lib/python2.7/dist-packages/dbus/bus.py", line 180, in activate_name_owner
    self.start_service_by_name(bus_name)
  File "/usr/lib/python2.7/dist-packages/dbus/bus.py", line 278, in start_service_by_name
    'su', (bus_name, flags)))
  File "/usr/lib/python2.7/dist-packages/dbus/connection.py", line 651, in call_blocking
    message, timeout)
DBusException: org.freedesktop.DBus.Error.ServiceUnknown: The name org.google.code.rabbitvcs.RabbitVCS.Checker was not provided by any .service files
`
Some sugest?

**Comment by jaheme127** [comment]
I found these kinds of errors in my logs as well, but I'm not sure when it happened... I still have several issues with rabbitvcs. 
I wonder why it still  refers to Python 2,7 in your case...

**Comment by nam20485** [comment]
What happened to `gtk3` branch? I don't see it. What version of python does the master branch use?

**Comment by monnerat** [comment]
The current master branch now contains code that is Python 2 & 3 compatible. Gtk2 is not supported anymore.
I cannot help you with Linux Mint packages, but you should now be able to run it with Nemo from a master branch snapshot.

**Comment by nam20485** [comment]
Yep. Thanks. I figured it had been merged into master. Thanks for the reply, that's all I needed.

**Comment by moisesbites** [comment]
I tried to install following the instruction here https://github.com/rabbitvcs/rabbitvcs/blob/master/clients/nemo/README

Running as root:
`
Traceback (most recent call last):
  File "/usr/share/nemo-python/extensions/RabbitVCS.py", line 85, in <module>
    from rabbitvcs.util.helper import to_text, unquote_url
ImportError: cannot import name to_text
Traceback (most recent call last):
  File "/usr/share/nemo-python/extensions/nemo-folder-color-switcher.py", line 20, in <module>
    import os, gettext, locale, urllib.parse, collections
ImportError: No module named parse
Initializing nemo-image-converter extension
`

**Comment by monnerat** [comment]
Well, to_text() is defined at https://github.com/rabbitvcs/rabbitvcs/blob/38583fceea2b94b98bb9a4c5f8801902c0ef4c97/rabbitvcs/util/helper.py#L100 and the nemo extension works fine here (Fedora 29 64bit, nemo version 4.0.6 using python 3.7).
Thus I may imagine you have some old RabbitVCS version files in the module load path that take precedence on the new ones. Did you install ALL up-to-date rabbitvcs/* files? Did you clean old version and old cache files?

**Comment by moisesbites** [comment]
> Well, to_text() is defined at
> 
> https://github.com/rabbitvcs/rabbitvcs/blob/38583fceea2b94b98bb9a4c5f8801902c0ef4c97/rabbitvcs/util/helper.py#L100
> and the nemo extension works fine here (Fedora 29 64bit, nemo version 4.0.6 using python 3.7).
> Thus I may imagine you have some old RabbitVCS version files in the module load path that take precedence on the new ones. Did you install ALL up-to-date rabbitvcs/* files? Did you clean old version and old cache files?

Hello monnerat,

I cleaned the system. I did again the instalation. For "root" user, all is ok. But for "non-root" users, RabbitVCS is not working for Nemo. I have a Linux Mint 19.1 with Python 3.7.3 (in system) and kernel 4.15.0-54-generic.

![Captura de tela de 2019-07-17 14-47-04](https://user-images.githubusercontent.com/8934577/61402743-c338ee80-a8a1-11e9-9055-e5e65837c069.png)
![Captura de tela de 2019-07-17 14-46-11](https://user-images.githubusercontent.com/8934577/61402744-c3d18500-a8a1-11e9-8ab4-dc841c735f88.png)

**Comment by monnerat** [comment]
@moisesbites 
If it runs as root, this is surely a module search path problem for the non-root user.
I'm afraid I can't help you more as I don't know anything about Linux Mint.

**Comment by rehierl** [comment]
I'd also love to use RabbitVCS on Linux Mint 19.1 with Nemo 4.0.6. Unfortunately, the synaptic package manager would only allow me to install v0.16, which is outdated since it's 5 years old. There are also only the following four packages listed: "rabbitvcs-cli", "-core", "-gedit", "-nautilus".

The [installation methods](http://wiki.rabbitvcs.org/wiki/install/ubuntu) didn't work: (1) The ppa instructions are just as outdated. I guess that would be the source packages for the package manager? - (2) The tarball link still points to code.google.com, which gets redirected to this repository. - (3) So I ended up giving the manual instructions a try ... and failed.

Most of the requirements mentioned for [debian-based distros](https://github.com/rabbitvcs/rabbitvcs/blob/master/README.md) I can resolve, but not all: (1) I can't use the package manager to install "python-gtk3" since that package isn't listed. ("sudo apt-get install python-gtk3" executed from within a termial will fail as well). - (2) There also is no "python-gtkspell" package listed. However, I could install a "python-gtkspellcheck" or a "python3-gtkspellcheck" package. - (3) There also is no "tkinter" package listed.

Despite these missing packages I did execute "sudo python setup.py install --install-layout=deb >output.txt". Since the generated output file didn't seem to list any issues, I continued to tinker around with no success.

I did follow the instructions mentioned in the [nemo readme](https://github.com/rabbitvcs/rabbitvcs/tree/master/clients/nemo). By the way, the "nemo-python" requirement should be changed into "python-nemo" since that seems to refer to the current package (i.e. nemo-python is v3.6 and python-nemo is v4.0). I also assume that "you may need to create python/python-extensions" refers to "/usr/share/pyhton/python-extensions"?. Even though I even restarted my system after these changes, it won't work.

From what I can tell, despite having no Python skills whatsoever, Nemo can't seem to recognize RabbitCVS as an extension. It won't get listed as an extension and also not as an "Action" in Nemo's "Plugins" overview. However, it will get listed as a "Script" if I copy "RabbitVCS.py" into "~/.local/share/nemo/scripts". But that seems to have no effect.

So yeah, I'm stuck since I lack the knowledge to get it working.

However, I'd be willing to give it another try if someone could find the time to update the instructions. Thanks in advance.

**Comment by CBtiger** [comment]
Hi 
I think to got it running. For Ubuntu/Xubuntu the instructions in the Readme.md are not fully correct and should be updated. I did the test on Xubuntu 18.04. 

Here is the step-by-step (the // stuff are comments, do not type them)

I removed all nemo and rabbitvcs installations
```
 sudo apt remove nemo* rabbitvcs*
 dpkg -l nemo* rabbitvcs*                                    // check if all were removed
```
Then fresh install
 sudo apt install nemo nemo-data nemo-python
 sudo apt install rabbitvcs-core rabbitvcs-cli         // cli might not be required
Download the `rabbitvcs/clients/nemo/` from this site (1 *.py file and one *.md). Copy the PY file to /usr/share/nemo-python/extension and set the permissions.
```
 sudo cp RabbitVCS.py  /usr/share/nemo-python/extension
 sudo chmod 755 /usr/share/nemo-python/extension/RabbitVCS.py
 ls -la
  rwxr-x-r-x   .... RabbitVCS.py               // if this is not right the extension will only work with sudo!
```

Now the sequence to install it in nemo. Here the Readme misses the sudo's
```
 sudo nemo  -q                                      // quite nemo and allow to find the new extension
 pgrep -f service.py | xargs sudo kill        // kills the rabbitvcs checkerservice
 nohup sudo nemo > /dev/null &             // restart and block errors. 
```
This will start nemo in elevated mode (sudo mode). As written in the posts before the extension should work. Make sure the config has ownership to you. For the user 'tiger' that would b
```
  ll ~/.config/rabbitvcs                                       // check you are owner
  sudo chown -R tiger:tiger ./config/rabbitvcs    // do not forget the recursive argument
```

I logged out and in. Then I could observer that the config file ./config/rabbitvcs/settings.conf contained some settings. Also the Plugin was listed in the normal mode when starting nemo. Also I could see icons when selecting a SVN subdirectory.

It would be nice if someone could confirm that this works on a rather fresh install as well. Maybe there are other dependencies that I already had installed. 


Optional infos:
There is a major problem right now with Python2.x and Python3.x. Whereever you take the file from, be prepared that you got the wrong version. Unfortunately its not in the filenames. The python3 version will be automatically converted from the python2.x. Seeing classes is a good sign its python2. 

I had been installing the PPA embrosyn/cinnamon because it featured a nemo-rabbitvcs before. But it failed to install with apt. That nemo was newer than 3.65 (which is current Ubuntu 18.04) and neeed pyhton3. Apt got blocked by having both packets in the same repro and could not resolve it. Now after going though the fully manual install, maye there is a chance to install nemo with all its very nice other extensions except nemo-rabbitvcs and just include the python3 version of the pyhton script manually.

**Comment by CBtiger** [comment]
Now I also got a standard Nemo4.0 working with RabbitVCS in Bionic.

I think the key is that 
- you need to isntall all dependencies as python3
- you need to install rabbitvcs as python3

Example
 sudo apt install python3-svn python3-gtkspellchecker
 python3 setup.py install --install-layout=deb 

I used for rabbitvcs and the nemo extension the lasted master download.

RabbitVCS: V17.01+
NemoRabbitVCS Extension: V17.01+
Nemo: 4.06

Works great. Only open item are the overlay icons (Emblems). They are tiny spots and I really like to get them like on their homepage  [http://wiki.rabbitvcs.org/wiki/about/screenshots](url) (see Issue #110)

**Comment by rehierl** [comment]
@CBtiger - Thanks to your feedback I did get it to work, including the emblems. Try to start Nemo from a terminal and see if it will report any missing packages (e.g. "tkinter" or "python3-tk").

To recap, the installation steps are as follows:

## RabbitVCS setup

(1) Download and unzip RabbitVCS from this repository.

(2) From within the unzipped `rabbitvcs-master` folder execute the following command.

- `sudo python3 setup.py install --install-layout=deb >output.txt`
- it is important to use "python3", not simply "python" (thanks CBtiger)
- check the output text file for any issues
- at the very end it should state that `./build/lib/rabbitvcs/buildinfo.py`
  was copied into `/usr/lib/python3/dist-packages/rabbitvcs`

(3) Check the contents of that rabbitvcs folder just to see if there is
something unusual (e.g. empty except for the copied buildinfo.py file).

## Nemo integration

(4) Copy `rabbitvcs-master/clients/nemo/RabbitVCS.py` file into `/usr/share/nemo-python/extensions` and make sure it has the execute permissions set (thanks CBtiger).

- from within /usr/share/nemo-python/extensions run
- `sudo chmod 755 RabbitVCS.py`

(5) Nemo should now list RabbitVCS as an extension. (Restart nemo if necessary).

## Test that it runs properly

(6) From within Nemo right-click into a folder and select `context-menu / RabbitVCS Git / Settings`.
Check that RabbitVCS also logs to the console (i.e. Type = Console or Both) and change the
log level to "Info".

(7) Shut down all Nemo instances and open Nemo from within a terminal. The terminal
window should display the names of any missing packages.

(8) Use the terminal's Nemo instance to create a new folder, move into it and
try to run `context-menu / RabbitVCS Git / Initialize Repository`. If that did
work, a window should pop open which should report success.

(9) If the initialization failed, check the terminal for any messages regarding
missing python modules. I had to use the synaptic package manager to install
the following two modules:

- message about "dulwich" not found - install: "python3-dulwich"
- message about "tkinter" not found - install: "python3-tk"
- if any other modules are missing, make sure to install the corresponding "python3-xxx" modules
- also keep in mind that packages could be named differently (e.g. "tkinter" -> "python3-tk")

**Comment by CBtiger** [comment]
Great that it worked for you. Maybe its helpful to add the versions and sources you have used.

All you checks were Ok I do not get errors. Still my embelms are not ok (see image in #101). So yours are not looking like that (I guess the difference is you are using cinnamon and not xfce)?

**Comment by rehierl** [comment]
> Maybe its helpful to add the versions and sources you have used.

RabbitVCS v0.17.1
- I did download and use the RabbitVCS sources from this repository.
- i.e. the current commit 5207510 from 2019-07-23

My system is
- Linux Mint 19.1 Cinnamon
- Nemo 4.0.6

> .. the overlay icons (Emblems). They are tiny spots ..

Yes, I do use the Cinnamon desktop, but that doesn't seem to be the issue since the overlay icons are indeed rather tiny. I agree that, compared to the size of the folder/file icons the emblems could be a little larger.

At my default zoom level, I can determine the emblem colors - but not as easily as it should be since I do have to focus a little. However, I have quite the difficulty to visually determine the symbols embedded into them (i.e. exclamation mark, check mark, ..). But that isn't too much of a problem since, in order to zoom in, I only have to use Ctrl+Mouse-Wheel. -- Yes, the emblems do scale with the folder/file icons.

Currently, they seem occupy 5-10% of the default icon size, where they should probably use maybe 25-50%.

**Comment by jaheme127** [comment]
May be someone can put the installation description to some promenint place? There are some hint, which took me long without looking in this thread?

**Comment by DMGPage** [comment]
Solution works for me too, but I have strange problem of viewing logs. I can see all history, but I can't write anything in search box. Can't even put cursor in that input field. Everything else seems work good.

**Comment by dtrunk90** [comment]
> write any

I'm facing the same problem on Ubuntu 20.04 LTS.

**Comment by SaiyanRiku** [comment]
Hello, for people that are looking for, i made a simple script to easily build a package with Nemo.
https://github.com/Jet1oeil/rabbitvcs-package 

I tested it on Linux LMDE 4 with nemo 4.6.5.

Hope this can help for people.

**Comment by elitegit** [comment]
Don't forget to install nemo-python

sudo apt-get install nemo-python

**Comment by csiefer** [comment]
Hi, great work ... 
but I get this:

Ubuntu 22.04.5 LTS
Python 3.10.12
nemo 5.2.4
nemo-python version 5.2.0-2

File "/usr/share/nemo-python/extensions/RabbitVCS.py", line 48
    except Exception, ex:
           ^^^^^^^^^^^^^
SyntaxError: multiple exception types must be parenthesized
ERROR SET META:15 Setting attribute metadata::nemo-list-view-sort-column not supported
ERROR SET META:15 Setting attribute metadata::nemo-list-view-sort-reversed not supported
ERROR SET META:15 Setting attribute metadata::nemo-list-view-zoom-level not supported

**Comment by monnerat** [comment]
> except Exception, ex:

This line has been changed almost 9 years ago: https://github.com/rabbitvcs/rabbitvcs/commit/7df3cd2f17675f47330700689a53e6100fd89fa3

Maybe it's time to upgrade...

**Comment by csiefer** [comment]
I got it from
https://github.com/embrosyn/rabbitvcs-nemo/archive/refs/heads/master.zip

**Comment by monnerat** [comment]
> I got it from https://github.com/embrosyn/rabbitvcs-nemo/archive/refs/heads/master.zip

So why don^t you report the problem there ?!?!?

IMO, you better follow the instructions in the current repo directory clients/nemo/README and avoid mixing repos.
There's very little if no chance you'll get support for a 3rd party repo here.

**Comment by csiefer** [comment]
Sorry for bothering, I'm just a noob.
Have a nice day
